File: tests/python/test_model_concat.py
"""Tests for model concatenation"""

import json

import numpy as np
import pytest

import treelite
from treelite.model_builder import (
    Metadata,
    ModelBuilder,
    PostProcessorFunc,
    TreeAnnotation,
)


@pytest.mark.parametrize("average_tree_output", [False, True])
def test_model_concat_with_tree_stump(average_tree_output):
    """Test model concatenation with a tree stump"""
    num_model_objs = 5

    def make_tree_stump():
        builder = ModelBuilder(
            threshold_type="float32",
            leaf_output_type="float32",
            metadata=Metadata(
                num_feature=2,
                task_type="kRegressor",
                average_tree_output=average_tree_output,
                num_target=1,
                num_class=[1],
                leaf_vector_shape=[1, 1],
            ),
            tree_annotation=TreeAnnotation(num_tree=1, target_id=[0], class_id=[0]),
            postprocessor=PostProcessorFunc(name="identity"),
            base_scores=[0.0],
        )
        builder.start_tree()
        builder.start_node(0)
        builder.numerical_test(
            feature_id=0,
            threshold=0.0,
            default_left=True,
            opname="<",
            left_child_key=1,
            right_child_key=2,
        )
        builder.end_node()
        builder.start_node(1)
        builder.leaf(1.0)
        builder.end_node()
        builder.start_node(2)
        builder.leaf(2.0)
        builder.end_node()
        builder.end_tree()
        return builder.commit()

    model_objs = [make_tree_stump() for _ in range(num_model_objs)]
    concatenated_model = treelite.Model.concatenate(model_objs)

    X = np.array([[-1.0, 0.0], [1.0, 0.0]], dtype=np.float32)
    pred = treelite.gtil.predict(concatenated_model, X)
    expected_pred = sum(treelite.gtil.predict(obj, X) for obj in model_objs)
    if average_tree_output:
        expected_pred /= len(model_objs)
    np.testing.assert_almost_equal(pred, expected_pred)

    ref_tree_stump = make_tree_stump()
    ref_tree_stump_json = json.loads(ref_tree_stump.dump_as_json(pretty_print=False))
    concatenated_model_json = json.loads(
        concatenated_model.dump_as_json(pretty_print=False)
    )
    for attr in [
        "num_feature",
        "task_type",
        "average_tree_output",
        "num_target",
        "num_class",
        "leaf_vector_shape",
        "postprocessor",
    ]:
        assert ref_tree_stump_json[attr] == concatenated_model_json[attr]
    for attr in ["target_id", "class_id"]:
        assert (
            ref_tree_stump_json[attr] * num_model_objs == concatenated_model_json[attr]
        )
    for tree in concatenated_model_json["trees"]:
        assert ref_tree_stump_json["trees"][0] == tree


def test_model_concat_with_mismatched_tree_types():
    """Test model concatenation with a tree stump"""

    def make_tree_stump(*, threshold_type, leaf_output_type):
        builder = ModelBuilder(
            threshold_type=threshold_type,
            leaf_output_type=leaf_output_type,
            metadata=Metadata(
                num_feature=2,
                task_type="kBinaryClf",
                average_tree_output=False,
                num_target=1,
                num_class=[1],
                leaf_vector_shape=[1, 1],
            ),
            tree_annotation=TreeAnnotation(num_tree=1, target_id=[0], class_id=[0]),
            postprocessor=PostProcessorFunc(name="sigmoid"),
            base_scores=[0.0],
        )
        builder.start_tree()
        builder.start_node(0)
        builder.numerical_test(
            feature_id=0,
            threshold=0.0,
            default_left=True,
            opname="<",
            left_child_key=1,
            right_child_key=2,
        )
        builder.end_node()
        builder.start_node(1)
        builder.leaf(1.0)
        builder.end_node()
        builder.start_node(2)
        builder.leaf(2.0)
        builder.end_node()
        builder.end_tree()
        return builder.commit()

    model_objs = [
        make_tree_stump(threshold_type="float32", leaf_output_type="float32"),
        make_tree_stump(threshold_type="float64", leaf_output_type="float64"),
    ]
    with pytest.raises(
        treelite.TreeliteError,
        match=r"Model object .* has a different type than the first model object",
    ):
        treelite.Model.concatenate(model_objs)

File: tests/cpp/test_model_concat.cc
/*!
 * Copyright (c) 2021-2023 by Contributors
 * \file test_model_concat.cc
 * \author Hyunsu Cho
 * \brief C++ tests for model concatenation
 */
#include <gtest/gtest.h>
#include <treelite/model_builder.h>
#include <treelite/tree.h>

#include <algorithm>
#include <memory>
#include <sstream>
#include <variant>
#include <vector>

namespace {

inline void TestRoundTrip(treelite::Model* model) {
  // Test round trip with in-memory serialization
  std::ostringstream oss;
  oss.exceptions(std::ios::failbit | std::ios::badbit);
  model->SerializeToStream(oss);

  std::istringstream iss(oss.str());
  iss.exceptions(std::ios::failbit | std::ios::badbit);
  std::unique_ptr<treelite::Model> received_model = treelite::Model::DeserializeFromStream(iss);

  // Use ASSERT_TRUE, since ASSERT_EQ will dump all the raw bytes into a string, potentially
  // causing an OOM error
  ASSERT_TRUE(model->DumpAsJSON(false) == received_model->DumpAsJSON(false));
}

}  // anonymous namespace

namespace treelite {

TEST(ModelConcatenation, TreeStump) {
  std::vector<std::unique_ptr<Model>> model_objs;
  constexpr int kNumModelObjs = 5;

  for (int i = 0; i < kNumModelObjs; ++i) {
    auto builder = model_builder::GetModelBuilder(TypeInfo::kFloat32, TypeInfo::kFloat32,
        model_builder::Metadata{2, TaskType::kRegressor, false, 1, {1}, {1, 1}},
        model_builder::TreeAnnotation{1, {0}, {0}}, model_builder::PostProcessorFunc{"identity"},
        {0.0});
    builder->StartTree();
    builder->StartNode(0);
    builder->NumericalTest(0, 0.0, true, Operator::kLT, 1, 2);
    builder->EndNode();
    builder->StartNode(1);
    builder->LeafScalar(1.0);
    builder->EndNode();
    builder->StartNode(2);
    builder->LeafScalar(2.0);
    builder->EndNode();
    builder->EndTree();
    model_objs.push_back(builder->CommitModel());
  }

  std::vector<Model const*> model_obj_refs;
  std::transform(model_objs.begin(), model_objs.end(), std::back_inserter(model_obj_refs),
      [](auto const& obj) { return obj.get(); });

  std::unique_ptr<Model> concatenated_model = ConcatenateModelObjects(model_obj_refs);
  ASSERT_EQ(concatenated_model->GetNumTree(), kNumModelObjs);
  EXPECT_EQ(concatenated_model->GetThresholdType(), TypeInfo::kFloat32);
  EXPECT_EQ(concatenated_model->GetLeafOutputType(), TypeInfo::kFloat32);
  EXPECT_TRUE(concatenated_model->target_id
              == ContiguousArray<std::int32_t>(std::vector<std::int32_t>(kNumModelObjs, 0)));
  EXPECT_TRUE(concatenated_model->class_id
              == ContiguousArray<std::int32_t>(std::vector<std::int32_t>(kNumModelObjs, 0)));
  TestRoundTrip(concatenated_model.get());
  auto& trees = std::get<ModelPreset<float, float>>(concatenated_model->variant_).trees;
  for (int i = 0; i < kNumModelObjs; ++i) {
    auto const& tree = trees[i];
    EXPECT_FALSE(tree.IsLeaf(0));
    EXPECT_TRUE(tree.IsLeaf(1));
    EXPECT_TRUE(tree.IsLeaf(2));
    EXPECT_EQ(tree.NodeType(0), TreeNodeType::kNumericalTestNode);
    EXPECT_EQ(tree.SplitIndex(0), 0);
    EXPECT_EQ(tree.Threshold(0), 0.0f);
    EXPECT_EQ(tree.LeftChild(0), 1);
    EXPECT_EQ(tree.RightChild(0), 2);
    EXPECT_EQ(tree.LeafValue(1), 1.0f);
    EXPECT_EQ(tree.LeafValue(2), 2.0f);
  }
}

TEST(ModelConcatenation, MismatchedTreeType) {
  std::vector<std::unique_ptr<Model>> model_objs;

  auto builder = model_builder::GetModelBuilder(TypeInfo::kFloat32, TypeInfo::kFloat32,
      model_builder::Metadata{2, TaskType::kRegressor, false, 1, {1}, {1, 1}},
      model_builder::TreeAnnotation{1, {0}, {0}}, model_builder::PostProcessorFunc{"identity"},
      {0.0});
  builder->StartTree();
  builder->StartNode(0);
  builder->LeafScalar(1.0);
  builder->EndNode();
  builder->EndTree();
  model_objs.push_back(builder->CommitModel());

  auto builder2 = model_builder::GetModelBuilder(TypeInfo::kFloat64, TypeInfo::kFloat64,
      model_builder::Metadata{2, TaskType::kRegressor, false, 1, {1}, {1, 1}},
      model_builder::TreeAnnotation{1, {0}, {0}}, model_builder::PostProcessorFunc{"identity"},
      {0.0});
  builder2->StartTree();
  builder2->StartNode(0);
  builder2->LeafScalar(1.0);
  builder2->EndNode();
  builder2->EndTree();
  model_objs.push_back(builder2->CommitModel());

  std::vector<Model const*> model_obj_refs;
  std::transform(model_objs.begin(), model_objs.end(), std::back_inserter(model_obj_refs),
      [](auto const& obj) { return obj.get(); });
  ASSERT_THROW(ConcatenateModelObjects(model_obj_refs), treelite::Error);
}

}  // namespace treelite
